cmake/Signing.cmake

```diff
@@ -0,0 +1,56 @@
+#
+#  Copyright (c) 2023, Intel Corporation
+#
+#  SPDX-License-Identifier: BSD-3-Clause
+
+#
+# ispc Signing.cmake
+#
+# Set of macros to sign build artifacts
+
+# Define a global property for tracking targets requiring signing
+define_property(GLOBAL PROPERTY ISPCSignTargetsList
+    BRIEF_DOCS "Tracks the list of build targets requiring signing"
+    FULL_DOCS "Tracks the list of build targets requiring signing"
+)
+
+# Macro to add a target to the list of targets requiring signing
+macro(target_sign TARGET)
+    get_property(ISPCSignTargetsValue GLOBAL PROPERTY ISPCSignTargetsList)
+    list(APPEND ISPCSignTargetsValue ${TARGET})
+    set_property(GLOBAL PROPERTY ISPCSignTargetsList ${ISPCSignTargetsValue})
+endmacro()
+
+# Macro to sign the files based on the targets in the ISPCSignTargetsList
+# This macro retrieves the targets, constructs the command, and performs the signing
+macro(sign_files)
+    if (ISPC_SIGN_KEY)
+        # Determine the signing command based on the platform
+        if (WIN32)
+            set(ISPC_SIGN_COMMAND signtool.exe)
+            set(ISPC_SIGN_ARGS sign /fd sha256 /sha1 ${ISPC_SIGN_KEY} /tr http://timestamp.comodoca.com/rfc3161 /td sha256)
+        elseif (APPLE)
+            set(ISPC_SIGN_COMMAND codesign)
+            set(ISPC_SIGN_ARGS -fs "${ISPC_SIGN_KEY}" --strict --timestamp --options=runtime)
+        endif()
+
+        # Retrieve the list of targets requiring signing from the global property
+        get_property(ISPC_SIGN_TARGETS GLOBAL PROPERTY ISPCSignTargetsList)
+
+        # Construct the list of files to sign by retrieving the target file paths
+        set(ISPC_SIGN_FILES "")
+        foreach(sign_target ${ISPC_SIGN_TARGETS})
+            list(APPEND ISPC_SIGN_FILES $<TARGET_FILE:${sign_target}>)
+        endforeach()
+
+        add_custom_target(sign
+            COMMAND ${ISPC_SIGN_COMMAND} ${ISPC_SIGN_ARGS} ${ISPC_SIGN_FILES}
```

Comment:
Should the sign command depends on files it signs, i.e., `DEPENDS ${ISPC_SIGN_FILES}`?

Reply:
> Starting CMake 3.16 it's not needed anymore to set dependent targets explicitly. CMake will ensure that `${ISPC_SIGN_FILES}` will be available before `sign` target is built.

At the moment, we have `3.13`: https://github.com/ispc/ispc/blob/a8a1b8b3889807212ddb17bc967e36c0df154d67/CMakeLists.txt#L9

Reply:
> Starting CMake 3.16 it's not needed anymore to set dependent targets explicitly. CMake will ensure that `${ISPC_SIGN_FILES}` will be available before `sign` target is built.

On my side this doesn't work (no `ispc` built triggered by `sign`), e.g.:

``` $ cmake ../
$ make sign
Build target sign
```

Reply:
That's a valid point. As a separate change I think it's time to switch to CMkae 1.19 where ISPC support was introduced in CMake.

Reply:
I fixed it.